### DISCUSSION TITLE
How to go about debugging intermittent RTT output

### DISCUSSION THREAD

**Discussion by jebiga52** [majbthrd/DapperMime#13] [Q&A]
I have a Raspberry Pi Pico running your firmware connected to a RP2040 on a custom PCB that I designed. I'm able to reliably flash the RP2040 using the DapperMime firmware (thank you for your hard work!) however I only occasionally see RTT output from the RP2040 coming back to my computer. It seems like RTT only occasionally works when both the probe and debug target have been fully power cycled, and doesn't work after that.

I'm unsure of where the problem originates in my stack, is it perhaps the debug probe running your firmware, is it maybe my PCB or the code running on it, is it perhaps the tooling I'm using to send RTT messages (probe-run and defmt)?

If you have any thoughts on how/where to chase this down I would appreciate it. Thank you.

**Comment by majbthrd** [comment]
I'm tempted to make the analogy of someone calling up a bridge builder to seek tech support on a smartphone because the smartphone got delivered by a truck that traveled over a bridge that happened to be built by the bridge builder.

I'm not even familiar with what RTT is, but it is certainly a protocol several layers above the CMSIS-DAP implementation.

DapperMime is just a compilation of ARM's reference CMSIS-DAP implementation.  If there is something wrong with DapperMime, then there is likely something wrong with every CMSIS-DAP implementation based on ARM's reference CMSIS-DAP implementation.

Throwing together many unknown things at once and then trying to sort out what is wrong is never a good way IMHO.

Things you can try to narrow it down:

1) Procure/borrow a CMSIS-DAP adapter using the same ARM reference implementation.  This would allow you to discount that your PCB is at issue.
2) Procure/borrow an ARM debug adapter that is not CMSIS-DAP.  If you have the same problems on this other adapter, then your tooling or your code is at issue.  If this other adapter doesn't have the issue, then the tooling has an incompatibility with CMSIS-DAP.
3) Try a different target other than RP2040.  This might discount the possibility that your code (or your PCB) is at issue.
